src/pymor/models/iosys.py

```diff
@@ -624,6 +625,40 @@ def to_abcde_files(self, files_basename):
         if E is not None:
             _mmwrite(Path(files_basename + '.E'), E)
 
+    def slice(self, input_indices=None, output_indices=None):
+        """Returns an |LTIModel| with a subselection of inputs and outputs.
+
+        Parameters
+        ----------
+        input_indices
+            Indices corresponding to a subselection of inputs to be paired with all outputs
+            or selected ones. If `None`, all inputs are used, otherwise, an `iterable` containing
+            the indices of the selected inputs has to be passed. The order of the inputs depends
+            on the order of the indices in the `iterable`. It is possible to pass negative indices
+            to access the inputs counting backwards.
+        output_indices
+            Indices corresponding to a subselection of outputs to be paired with all inputs
+            or selected ones. If `None`, all outputs are used, otherwise, an `iterable` containing
+            the indices of the selected outputs has to be passed. The order of the outputs depends
+            on the order of the indices in the `iterable`. It is possible to pass negative indices
+            to access the outputs counting backwards.
+        """
+        if input_indices is None:
+            input_indices = list(range(self.dim_input))
+        if output_indices is None:
+            output_indices = list(range(self.dim_output))
+
+        assert all(isinstance(item, int) for item in chain(input_indices, output_indices))
+        assert all(-self.dim_input <= item < self.dim_input for item in input_indices), \
+            f'input_indices should be any integer value between {-self.dim_input} and {self.dim_input-1}'
+        assert all(-self.dim_output <= item < self.dim_output for item in output_indices), \
+            f'output_indices should be any integer value between {-self.dim_output} and {self.dim_output-1}'
+
+        m, p = len(input_indices), len(output_indices)
+        return self.with_(B=NumpyMatrixOperator(to_matrix(self.B)[:, input_indices].reshape(-1, m)),
+                          C=NumpyMatrixOperator(to_matrix(self.C)[output_indices, :].reshape(p, -1)),
+                          D=NumpyMatrixOperator(to_matrix(self.D)[output_indices, input_indices].reshape(p, m)))
```

Comment:
I guess, to allow parametric `B`, `C`, and `D`, we would need to use `ConcatenationOperators`, e.g.,

```python
self.B @ NumpyMatrixOperator(np.eye(self.dim_input)[:, input_indices])
```

Additionally, `pymor.algorithms.simplify.contract` could be called to do the multiplication in the case of non-parametric `Operators`.

Reply:
I'd agree with @pmli that this should be the fallback. This not only affects the parametric models but also models coming from an external pde solver. (It would be slightly better to create a `VectorArrayOperator` instead of a `NumpyMatrixOperator` as `contract` will convert it anyway.) One could still have this as a special case when `B`, `C` are `NumpyMatrixOperators`. But maybe it's not worth performance-wise to have extra code.